Change exception for overflow in ArrayBufferWriter

felipepessoto opened the pull request:
Continuation from https://github.com/dotnet/runtime/pull/1308

felipepessoto committed:
Fix #609 - Part 2
We throw a OutOfMemoryException instead of "Arithmetic operation resulted in an overflow"

felipepessoto commented:
https://github.com/dotnet/runtime/blob/9505b7314f291e3788cb5e557b0d62e4c1103433/src/libraries/Common/src/System/Buffers/ArrayBufferWriter.cs#L173-L176

This if is not clear to me. If the buffer.Length is 0, like when we just created the instance. It doesn't respect the sizeHint

ahsonkhan commented on src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.T.cs:
```diff
@@ -205,6 +205,19 @@ public void GetMemory_DefaultCtor(int sizeHint)
+        [PlatformSpecific(TestPlatforms.Windows | TestPlatforms.OSX)]
+        [ConditionalFact(nameof(IsX64))]
+        [OuterLoop]
+        public void GetMemory_ExceedMaximumBufferSize()
```
Does this test actually end up allocating a lot or taking a while? If it isn't resource intensive (which it doesn't look like to be), it could just be `[Fact]`.

felipepessoto replied:
We may also remove the OuterLoop?

ahsonkhan replied:
Yes.

felipepessoto committed:
Test attributes. We don't need to limit to 64bit process, since the memory is never allocated.

felipepessoto commented:
The test is failing in "netcoreapp5.0-OSX-Debug-x64-Mono_release-OSX.1013.Amd64.Open" environment.
@ahsonkhan, do you know what can cause this difference?

ahsonkhan commented:
I am not sure why that is failing, but it looks like it is mono specific:
https://helix.dot.net/api/2019-06-17/jobs/d839f076-ff49-437f-ae7b-77ff86928be6/workitems/System.Memory.Tests/console

```text
Starting:    System.Memory.Tests (parallel test collections = on, max threads = 4)
    System.Buffers.Tests.ArrayBufferWriterTests_String.GetMemory_ExceedMaximumBufferSize [FAIL]
      Assert.Throws() Failure
      Expected: typeof(System.OutOfMemoryException)
      Actual:   (No exception was thrown)
      Stack Trace:
        /_/src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.T.cs(217,0): at System.Buffers.Tests.ArrayBufferWriterTests`1[[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].GetMemory_ExceedMaximumBufferSize()
        /_/src/mono/netcore/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.cs(339,0): at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
    System.Buffers.Tests.ArrayBufferWriterTests_Byte.GetMemory_ExceedMaximumBufferSize [FAIL]
      Assert.Throws() Failure
      Expected: typeof(System.OutOfMemoryException)
      Actual:   (No exception was thrown)
      Stack Trace:
        /_/src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.T.cs(217,0): at System.Buffers.Tests.ArrayBufferWriterTests`1[[System.Byte, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].GetMemory_ExceedMaximumBufferSize()
        /_/src/mono/netcore/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.cs(339,0): at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
    System.Buffers.Tests.ArrayBufferWriterTests_Char.GetMemory_ExceedMaximumBufferSize [FAIL]
      Assert.Throws() Failure
      Expected: typeof(System.OutOfMemoryException)
      Actual:   (No exception was thrown)
      Stack Trace:
        /_/src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.T.cs(217,0): at System.Buffers.Tests.ArrayBufferWriterTests`1[[System.Char, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].GetMemory_ExceedMaximumBufferSize()
        /_/src/mono/netcore/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.cs(339,0): at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
  Finished:    System.Memory.Tests
```

@jkotas - do you know why we aren't throwing OOM on mono OSX here? Does it support more than int.MaxValue elements somehow?
```C#
var buffer = new byte[10];
Array.Resize<byte>(ref buffer, int.MaxValue);
```

cc @alexischr

ahsonkhan commented on src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.T.cs:
```diff
@@ -205,6 +205,18 @@ public void GetMemory_DefaultCtor(int sizeHint)
+        //       problems on Linux due to the way deferred memory allocation works. On Linux, the allocation can 
+        //       succeed even if there is not enough memory but then the test may get killed by the OOM killer at the 
+        //       time the memory is accessed which triggers the full memory allocation. 
+        [PlatformSpecific(TestPlatforms.Windows | TestPlatforms.OSX)]
```
You can remove this platform limitation along with the comment. It doesn't apply here.

felipepessoto committed:
Remove platform restriction

jkotas commented:
> do you know why we aren't throwing OOM on mono OSX here? Does it support more than int.MaxValue elements somehow?

Look for `PlatformDetection.IsNotIntMaxValueArrayIndexSupported`

ahsonkhan commented:
> https://github.com/dotnet/runtime/blob/9505b7314f291e3788cb5e557b0d62e4c1103433/src/libraries/Common/src/System/Buffers/ArrayBufferWriter.cs#L173-L176
> 
> This (pre-existing) **_if_** is not clear to me. If the buffer.Length is 0, like when we just created the instance. It doesn't respect the sizeHint

The `IBufferWriter<T>` interface requires that we return a buffer that is **at least** as big as `sizeHint`. We can return something bigger if that is most optimal:
https://docs.microsoft.com/en-us/dotnet/api/system.buffers.ibufferwriter-1.getspan?view=netcore-3.1#System_Buffers_IBufferWriter_1_GetSpan_System_Int32_

In this case, if a user keeps asking for a buffer passing sizeHint = 0, it doesn't make sense to start the "doubling" with 1 (then 2, 4, 8, ...). So, we cap the lower bound by 256, which seems to be a reasonable starting size for most operations.

felipepessoto commented:
> > do you know why we aren't throwing OOM on mono OSX here? Does it support more than int.MaxValue elements somehow?
> 
> Look for `PlatformDetection.IsNotIntMaxValueArrayIndexSupported`

In this case, should we explicitly throw an OutOfMemory, instead of setting newSize to int.MaxValue (I understand we only do that because of its side-effect is to throw an OOM Exception when we callArray.Resize. We may also need to do the same in my previous PR)

ahsonkhan commented:
Yep, looks like we need to disable it for platforms that support arrays larger than int.Max:
https://github.com/mono/mono/issues/15189#issuecomment-503429568

ahsonkhan commented:
> If we keep it the way it is now, in OSX Mono environment that method will return a Buffer of int.MaxValue size, instead of the size required, breaking the contract.

Sure, that makes sense. I am not sure what we generally do in the rest of the framework for this type of discrepancy, but given I was leaning towards having a custom exception message anyway, this works (https://github.com/dotnet/runtime/pull/1308#discussion_r369911976). Consider adding and calling a throwhelper method similar to:
https://github.com/dotnet/runtime/blob/d515c203410a50b2725400c969566c9f484a01ee/src/libraries/Common/src/System/Buffers/ArrayBufferWriter.cs#L195-L198

cc @stephentoub

felipepessoto commented:
@stephentoub, should we explicitly throw OOM? And change my previous PR (https://github.com/dotnet/runtime/pull/1308) to do the same?

Thanks

ahsonkhan commented:
> should we explicitly throw OOM?

For unblocking this change/PR, **let's do it**. I doubt we'll need to, but we can re-visit that decision later.

felipepessoto committed:
Explicitly throw the OutOfMemoryException instead of rely on Array.Resize

felipepessoto committed:
Move ThrowOutOfMemoryException method

ahsonkhan commented on src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.T.cs:
```diff
@@ -205,6 +205,13 @@ public void GetMemory_DefaultCtor(int sizeHint)
+        public void GetMemory_ExceedMaximumBufferSize()
+        {
+            var output = new ArrayBufferWriter<T>(256);
+            Assert.Throws<OutOfMemoryException>(() => output.GetMemory(int.MaxValue));
```
Are all the new branches of `CheckAndResizeBuffer` covered by this/other unit tests? Can you please verify by doing a quick code coverage run, locally?

felipepessoto replied:
Now it is :)

ahsonkhan commented on src/libraries/Common/src/System/Buffers/ArrayBufferWriter.cs:
```diff
@@ -186,5 +196,10 @@ private static void ThrowInvalidOperationException_AdvancedTooFar(int capacity)
+
+        private static void ThrowOutOfMemoryException(uint capacity)
+        {
+            throw new OutOfMemoryException(SR.Format(SR.BufferMaximumSizeExceeded, capacity));
```
Where is `BufferMaximumSizeExceeded` defined?

felipepessoto replied:
I just commited the resx file

felipepessoto committed:
Missing message, BufferMaximumSizeExceeded

felipepessoto committed:
Code Coverage

ahsonkhan commented on src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.T.cs:
```diff
@@ -205,6 +205,15 @@ public void GetMemory_DefaultCtor(int sizeHint)
+        {
+            var output = new ArrayBufferWriter<T>(int.MaxValue / 2 + 1);
+            output.Advance(int.MaxValue / 2 + 1);
+            output.GetMemory(1);//Validate when can't double the buffer size, but can grow by sizeHint
```
nit: Consider asserting that the returned `Memory<T>` is of the expected size.

ahsonkhan replied:
Also, typo in comment:
```suggestion
            output.GetMemory(1); // Validate we can't double the buffer size, but can grow by sizeHint
```

ahsonkhan commented on src/libraries/System.Memory/src/Resources/Strings.resx:
```diff
@@ -146,4 +147,7 @@
     <value>Cannot advance past the end of the buffer, which has a size of {0}.</value>
   </data>
+  <data name="BufferMaximumSizeExceeded" xml:space="preserve">
+    <value>Cannot allocate buffer of size {0}.</value>
```
nit: add `a`
```suggestion
    <value>Cannot allocate a buffer of size {0}.</value>
```

ahsonkhan reviewed (approved):
Some nits. Otherwise, looks good.

ahsonkhan commented:
> And change my previous PR (#1308) to do the same?

Yes please ;)

ahsonkhan commented on src/libraries/Common/src/System/Buffers/ArrayBufferWriter.cs:
```diff
@@ -186,5 +196,10 @@ private static void ThrowInvalidOperationException_AdvancedTooFar(int capacity)
             throw new InvalidOperationException(SR.Format(SR.BufferWriterAdvancedTooFar, capacity));
         }
+
+        private static void ThrowOutOfMemoryException(uint capacity)
```
Because `ArrayBufferWriter` is referenced both in System.Memory and in System.Text.Json, we need to add `BufferMaximumSizeExceeded` to the resx file in System.Text.Json as well.

For example, see `SR.BufferWriterAdvancedTooFar`:
https://github.com/dotnet/runtime/search?q=BufferWriterAdvancedTooFar&unscoped_q=BufferWriterAdvancedTooFar

https://github.com/dotnet/runtime/blob/4a804959512a76556add344d998f3fc9ca9513e8/src/libraries/System.Text.Json/src/Resources/Strings.resx#L309-L311

See the `ArrayBufferWriter.cs` file being included in that project:
https://github.com/dotnet/runtime/blob/4a804959512a76556add344d998f3fc9ca9513e8/src/libraries/System.Text.Json/src/System.Text.Json.csproj#L230-L232

It's used in the Utf8JsonWriter and in S.T.J.Tests (this internal source copy is needed for netstandard2.0):
https://github.com/dotnet/runtime/blob/81bf79fd9aa75305e55abe2f7e9ef3f60624a3a1/src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs#L46

felipepessoto committed:
Resource file

felipepessoto committed:
Missing Resx update

felipepessoto commented:
This test (https://github.com/dotnet/runtime/pull/32587/checks?check_run_id=530741887) is failing:

```
   at System.Buffers.ArrayBufferWriter`1[[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]..ctor(Int32 initialCapacity) in /_/src/libraries/Common/src/System/Buffers/ArrayBufferWriter.cs:line 47
   at System.Buffers.Tests.ArrayBufferWriterTests`1[[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].GetMemory_ExceedMaximumBufferSize() in /_/src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.T.cs:line 211
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) in /_/src/mono/netcore/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.cs:line 339
```

after I added this line in the unit test:

`new ArrayBufferWriter<T>(int.MaxValue / 2 + 1);`

I'll restrict it to 64bits environment.

felipepessoto committed:
Restrict test to 64bits environment

felipepessoto commented:
> > And change my previous PR (#1308) to do the same?
> 
> Yes please ;)

Done: https://github.com/dotnet/runtime/pull/34040
The build is failing, but it seems an existing problem

ahsonkhan commented on src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.T.cs:
```diff
@@ -205,6 +205,17 @@ public void GetMemory_DefaultCtor(int sizeHint)
             Assert.Equal(sizeHint <= 256 ? 256 : sizeHint, memory.Length);
         }
 
+        [ConditionalFact(nameof(IsX64))]
```
Since we are trying to avoid the OOM, don't run this on Linux.
https://github.com/dotnet/runtime/blob/d59f143221bfa500fe2b974827dcb47eb4ce327c/src/libraries/System.Text.Json/tests/Serialization/WriteValueTests.cs#L384-L388

ahsonkhan commented on src/libraries/System.Text.Json/tests/Resources/Strings.resx:
```diff
@@ -20278,4 +20278,7 @@ tiline\"another\" String\\"],"str":"\"\""}</value>
     <value>TG9yZW0gaXBzdW0gZG9sb3Igc2l0IGFtZXQsIGNvbnNlY3RldHVyIGFkaXBpc2NpbmcgZWxpdC4gQ3VyYWJpdHVyIHZpdmVycmEgbmVxdWUgYXQgZXJhdCBsYW9yZWV0LCBhdCBzYWdpdHRpcyBhcmN1IGVmZmljaXR1ci4gVXQgcHVsdmluYXIgZXJvcyBuZWMgb2RpbyBjdXJzdXMgZWxlaWZlbmQuIE1hZWNlbmFzIHZpdmVycmEgZWxlbWVudHVtIHBvcnR0aXRvci4gTnVsbGFtIG1pIHZlbGl0LCBtYWxlc3VhZGEgY29tbW9kbyB0cmlzdGlxdWUgZXUsIG1vbGxpcyBhIHZlbGl0LiBNb3JiaS4=</value>
   </data>
-</root>
+  <data name="BufferMaximumSizeExceeded" xml:space="preserve">
```
Just to confirm, was adding this to the tests project necessary? It probably is given we include the `ArrayBufferWriter.cs` file, but curios if you saw a build failure without it and then added it or added it preemptively.

felipepessoto replied:
Yes, it was failing: https://github.com/dotnet/runtime/pull/32587/commits/0f4bc7984ff461c22f714cd1068ef87c80cab509
https://github.com/dotnet/runtime/runs/530510806

ahsonkhan reviewed (approved):
Some test suggestions.

felipepessoto committed:
Simple test so that we have some test coverage in inner loop
Since we are trying to avoid the OOM, don't run the GetMemory_ExceedMaximumBufferSize on Linux

ahsonkhan commented on src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.T.cs:
```diff
@@ -205,6 +205,25 @@ public void GetMemory_DefaultCtor(int sizeHint)
+            Assert.Throws<OutOfMemoryException>(() => output.GetMemory(int.MaxValue));
+        }
+
+        [PlatformSpecific(TestPlatforms.Windows | TestPlatforms.OSX)]
```
nit: Add the comment explaining why we are leaving out linux. We have been doing that elsewhere as well to keep the reason inline with the test.

felipepessoto committed:
Test comment

ahsonkhan commented:
> I'll restrict it to 64bits environment.

I am getting OOM even on my local machine. The test is taking up 12+ GB of memory to run (even if we don't call `output.GetMemory(int.MaxValue)`). Even for outerloop that's too much.
Please re-write it to not be so resource intensive.

The problem is the `ArrayBufferWriterTests_String` class is allocating a lot (~16 GB).

Rather adding the test to the generic base class, make it specific to byte only and just add it to the `ArrayBufferWriterTests_Byte` tests (and remove it from the generic test class). That way you only allocate ~2 GB.
https://github.com/dotnet/runtime/blob/4040e6069eefe6441e1d9c31f9ec2cd8a133b946/src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.Byte.cs#L11-L12
```C#
// NOTE: GetMemory_ExceedMaximumBufferSize test is constrained to run on Windows and MacOSX because it causes
//       problems on Linux due to the way deferred memory allocation works. On Linux, the allocation can
//       succeed even if there is not enough memory but then the test may get killed by the OOM killer at the
//       time the memory is accessed which triggers the full memory allocation.
[PlatformSpecific(TestPlatforms.Windows | TestPlatforms.OSX)]
[ConditionalFact(nameof(IsX64))]
[OuterLoop]
public void GetMemory_ExceedMaximumBufferSize()
{
    var output = new ArrayBufferWriter<byte>(int.MaxValue / 2 + 1);
    output.Advance(int.MaxValue / 2 + 1);
    Memory<byte> memory = output.GetMemory(1); // Validate we can't double the buffer size, but can grow by sizeHint
    Assert.Equal(1, memory.Length);
    Assert.Throws<OutOfMemoryException>(() => output.GetMemory(int.MaxValue));
}
```

felipepessoto committed:
Move OutOfMemory test to byte specific test

ahsonkhan commented:
CI failure is unrelated: https://github.com/dotnet/runtime/issues/34087
